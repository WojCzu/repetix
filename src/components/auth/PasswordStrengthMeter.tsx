import { cn } from "@/lib/utils";

interface PasswordStrengthMeterProps {
  password: string;
  className?: string;
}

export function PasswordStrengthMeter({ password, className }: PasswordStrengthMeterProps) {
  const calculateStrength = (password: string): number => {
    if (!password) return 0;

    let strength = 0;

    // Length check
    if (password.length >= 8) strength += 1;

    // Character variety checks
    if (/[A-Z]/.test(password)) strength += 1;
    if (/[a-z]/.test(password)) strength += 1;
    if (/[0-9]/.test(password)) strength += 1;
    if (/[^A-Za-z0-9]/.test(password)) strength += 1;

    return Math.min(strength, 5);
  };

  const strength = calculateStrength(password);
  const strengthText = ["Very Weak", "Weak", "Fair", "Good", "Strong"][strength - 1] || "No Password";
  const strengthColor = {
    0: "bg-muted",
    1: "bg-destructive",
    2: "bg-orange-500",
    3: "bg-yellow-500",
    4: "bg-green-500",
    5: "bg-green-600",
  }[strength];

  return (
    <div className={cn("space-y-2", className)}>
      <div className="flex h-2 w-full gap-1">
        {Array.from({ length: 5 }).map((_, i) => (
          <div
            key={i}
            className={cn("h-full w-full rounded-full transition-colors", i < strength ? strengthColor : "bg-muted")}
          />
        ))}
      </div>
      <p className="text-xs text-muted-foreground">{strengthText}</p>
    </div>
  );
}
