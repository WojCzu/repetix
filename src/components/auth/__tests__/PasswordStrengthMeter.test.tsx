import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { PasswordStrengthMeter } from "../PasswordStrengthMeter";

describe("PasswordStrengthMeter", () => {
  // Tests for strength calculation
  describe("calculateStrength function", () => {
    const testCases = [
      { password: "", expected: 0, description: "empty password" },
      { password: "short", expected: 1, description: "short password" },
      { password: "password12345", expected: 3, description: "long password with lowercase and numbers" },
      { password: "Password12345", expected: 4, description: "password with uppercase, lowercase, and numbers" },
      {
        password: "Password!2345",
        expected: 5,
        description: "password with uppercase, lowercase, numbers, and special chars",
      },
      { password: "P@ssw0rd!", expected: 5, description: "strong password with all criteria" },
      { password: "P@ss1", expected: 4, description: "short password with all character types" },
    ];

    testCases.forEach(({ password, expected, description }) => {
      it(`should return strength ${expected} for ${description}`, () => {
        const { container } = render(<PasswordStrengthMeter password={password} />);

        // Count filled indicators
        const filledIndicators = container.querySelectorAll('[class*="bg-"]:not([class*="bg-muted"])').length;

        expect(filledIndicators).toBe(expected);
      });
    });
  });

  // Tests for visual representation
  describe("visual representation", () => {
    it("should display correct text based on password strength", () => {
      const strengthTexts = [
        { password: "", expected: "No Password" },
        { password: "pass", expected: "Very Weak" },
        { password: "password12345", expected: "Fair" },
        { password: "Password12345", expected: "Good" },
        { password: "Password!2345", expected: "Strong" },
        { password: "P@ssw0rd!", expected: "Strong" },
      ];

      strengthTexts.forEach(({ password, expected }) => {
        const { unmount } = render(<PasswordStrengthMeter password={password} />);
        expect(screen.getByText(expected)).toBeInTheDocument();
        unmount();
      });
    });

    it("should apply correct colors based on strength", () => {
      const strengthColors = [
        { password: "a", strengthClass: "bg-destructive" },
        { password: "Password1", strengthClass: "bg-green-500" },
        { password: "Password1!", strengthClass: "bg-green-600" },
        { password: "P@ssw0rd!", strengthClass: "bg-green-600" },
      ];

      strengthColors.forEach(({ password, strengthClass }) => {
        const { container, unmount } = render(<PasswordStrengthMeter password={password} />);

        // At least one indicator should have the expected color
        const hasColorClass = Array.from(container.querySelectorAll(".h-full.w-full")).some((element) =>
          element.className.includes(strengthClass)
        );
        expect(hasColorClass).toBe(true);

        unmount();
      });
    });

    it("should use bg-muted for unfilled indicators", () => {
      const { container } = render(<PasswordStrengthMeter password="a" />);

      // For strength 1, 4 indicators should be muted
      const mutedIndicators = container.querySelectorAll(".bg-muted").length;
      expect(mutedIndicators).toBeGreaterThan(0);
    });
  });

  // Test component props
  it("should apply custom className", () => {
    const customClass = "custom-test-class";
    const { container } = render(<PasswordStrengthMeter password="test" className={customClass} />);

    expect(container.firstChild).toHaveClass(customClass);
  });

  // Test the correct number of indicators
  it("should render 5 strength indicators", () => {
    const { container } = render(<PasswordStrengthMeter password="test" />);

    // Select only the strength meter indicators
    const indicators = container.querySelectorAll(".flex.h-2.w-full.gap-1 > div");
    expect(indicators.length).toBe(5);
  });

  // Snapshot test
  it("should match snapshot for different strength levels", () => {
    ["", "a", "password", "Password1", "P@ssw0rd!"].forEach((password) => {
      const { container, unmount } = render(<PasswordStrengthMeter password={password} />);
      expect(container.firstChild).toMatchSnapshot();
      unmount();
    });
  });
});
