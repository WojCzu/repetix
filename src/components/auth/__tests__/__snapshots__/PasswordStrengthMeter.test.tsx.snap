// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PasswordStrengthMeter > should match snapshot for different strength levels 1`] = `
<div
  class="space-y-2"
>
  <div
    class="flex h-2 w-full gap-1"
  >
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
  </div>
  <p
    class="text-xs text-muted-foreground"
  >
    No Password
  </p>
</div>
`;

exports[`PasswordStrengthMeter > should match snapshot for different strength levels 2`] = `
<div
  class="space-y-2"
>
  <div
    class="flex h-2 w-full gap-1"
  >
    <div
      class="h-full w-full rounded-full transition-colors bg-destructive"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
  </div>
  <p
    class="text-xs text-muted-foreground"
  >
    Very Weak
  </p>
</div>
`;

exports[`PasswordStrengthMeter > should match snapshot for different strength levels 3`] = `
<div
  class="space-y-2"
>
  <div
    class="flex h-2 w-full gap-1"
  >
    <div
      class="h-full w-full rounded-full transition-colors bg-orange-500"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-orange-500"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
  </div>
  <p
    class="text-xs text-muted-foreground"
  >
    Weak
  </p>
</div>
`;

exports[`PasswordStrengthMeter > should match snapshot for different strength levels 4`] = `
<div
  class="space-y-2"
>
  <div
    class="flex h-2 w-full gap-1"
  >
    <div
      class="h-full w-full rounded-full transition-colors bg-green-500"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-green-500"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-green-500"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-green-500"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-muted"
    />
  </div>
  <p
    class="text-xs text-muted-foreground"
  >
    Good
  </p>
</div>
`;

exports[`PasswordStrengthMeter > should match snapshot for different strength levels 5`] = `
<div
  class="space-y-2"
>
  <div
    class="flex h-2 w-full gap-1"
  >
    <div
      class="h-full w-full rounded-full transition-colors bg-green-600"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-green-600"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-green-600"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-green-600"
    />
    <div
      class="h-full w-full rounded-full transition-colors bg-green-600"
    />
  </div>
  <p
    class="text-xs text-muted-foreground"
  >
    Strong
  </p>
</div>
`;
